sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample09_MockingEventHubPublisher.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs.Producer;
+using Moq;
+using NUnit.Framework;
+
+namespace Azure.Messaging.EventHubs.Tests.Snippets
+{
+    [TestFixture]
+    public class Sample09_MockingEventHubPublisher
+    {
+        /// <summary>
+        ///     Performs basic unit test validation of the contained snippet.
+        /// </summary>
+        [Test]
+        public async Task EventBatch()
+        {
+            #region Snippet:EventHubs_Sample09_EventBatchMock
+
+            var size = 1024;
+            var store = new List<EventData> { new EventData(Array.Empty<byte>()), new EventData(Array.Empty<byte>()) };
+            var mockResponse = new Mock<Response>();
+            Response<EventDataBatch> response = Response.FromValue(EventHubsModelFactory.EventDataBatch(size, store), mockResponse.Object);
+
+            var mockEventHubProducerClient = new Mock<EventHubProducerClient>();
+
+            mockEventHubProducerClient.Setup(c => c.CreateBatchAsync(It.IsAny<CancellationToken>())).ReturnsAsync(response);
+
+            var producer = mockEventHubProducerClient.Object;
+
+            try
+            {
+                using var eventBatch = await producer.CreateBatchAsync();
+
+                var eventBody = new BinaryData("This is an event body");
+                var eventData = new EventData(eventBody);
+
+                if (!eventBatch.TryAdd(eventData))
+                {
+                    throw new Exception($"The event could not be added.");
+                }
+            }
+            finally
+            {
+                await producer.CloseAsync();
+            }
+
+            #endregion
```

Comment:
We probably want to show validation like developers would be doing in a test.   Creating a set of events up front, adding them to the batch, and then verifying that each event is in the batch would be helpful.   Something like:

```csharp

var expectedEvents = GenerateEventList( ... );

// Create the producer and batch the way we're doing now...

// Let's ensure that we see the Send call.

mockEventHubProducerClient
    .Setup(c => c.SendAsync(mockEventBatch, It.IsAny<CancellationToken>())
    .Returns(Task.CompletedTask)
    .Verifiable();

foreach (var eventData in expectedEvents)
{
     if (!eventBatch.TryAdd(eventData))
     {
          // My callback rejected the event....
     }
}

await producer.SendAsync(eventBatch, CancellationToken.None);

// Did the events that I think that I added to the batch actually make it into the
// batch?

for (var index = 0; index < expectedEvents.Cout; ++index)
{
    Assert.AreEqual(expectedEvents[index].EventBody.ToString(), eventStore[index].EventBody.ToString());
}

mockEventHubProducerClient.VerifyAll();
```

Reply:
I'm deliberately using the `Assert.AreEqual` form here, so that it is more easily translatable for developers using xUnit or MSTest or another test framework.